### PR TITLE
KAFKA-13676: Commit successfully processed tasks on error

### PR DESCRIPTION
**Pull request by wcarlson5**
When we hit an exception when processing tasks we should save the work we have done so far.
This will only be relevant with ALOS and EOS-v1, not EOS-v2. It will actually reduce the number of duplicated record in ALOS because we will not be successfully processing tasks successfully more than once in many cases.

This is currently enabled only for named topologies.

The behavior was rather throughly tested. There is a `EmitOnChangeIntegrationTest` that makes sure if there is an exception in a task, after replacing the thread and if the exception does not throw again the record would still be processed. I just added another task into the test such that even if the other task keep throwing exceptions and hence never complete processing its record, this new task that could process successfully would be able to make progress and commit its offset.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by wcarlson5** [comment]
@guozhangwang @ableegoldman This is ready for a pass when you have a sec

File: streams/src/test/java/org/apache/kafka/streams/integration/EmitOnChangeIntegrationTest.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/integration/EmitOnChangeIntegrationTest.java]: We can simplify it to `stream().peek(throw).to()` without materializing a store.

File: streams/src/test/java/org/apache/kafka/streams/integration/EmitOnChangeIntegrationTest.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/integration/EmitOnChangeIntegrationTest.java:237]: Should we also check that `outputTopic` never sees the record `1 -> A` since that task kept throwing exception?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java:78]: If an exception throws here, it means we should have not reached line 93 below and hence this task should not have been added to the set, and hence we do not need to ever call `removeTaskFromCuccessfullyProcessedBeforeClosing` right?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java]: The check on the processingMode happens in two places, which seems redundant to me: e.g. if it is not ALOS, then the `successfullyProcessed()` should always be empty here. I think we can simply this line to only consider the second condition.

**Review by guozhangwang — COMMENTED** [review]
I left some detailed comments in the PR.

One meta question though is that, committing offsets is not a free operation as it incurs one round-trip from the brokers, and we are trying to batch all tasks in a single call. 

Say if we have 100 tasks, and the 2nd task in the list keep throwing an error while the 1st task could be processed successfully, we would end up keep consuming the 1st task only and leaving the rest 98 tasks untouched which hurts the performance a lot.

With that in mind, I think we need to consider this PR along with e.g. sorting the tasks to put the error-ed ones to the back of the list. So I'd suggest we only merge this PR after the other work is done. WDYT?

File: streams/src/test/java/org/apache/kafka/streams/integration/EmitOnChangeIntegrationTest.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/integration/EmitOnChangeIntegrationTest.java:245]: Also I think by just checking that 1->A and 1->B are there we do not guarantee there's no duplicates due to re-processing right? I think we should check that the offset on the input topic can be committed and also there's no duplicates in the output.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java]: Can be `AT_LEAST_ONCE` or `EXACTLY_ONCE_V1`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java
Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java:78]: Only if we hit the error in the first set of processing. We could have already processed it once this iteration and are coming back todo it again. I ran into this issue while testing and I m pretty sure that is why it is coming up again.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java
Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java]: sounds good. I have no issues removing this check

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java
Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java]: How about `!EXACTLY_ONCE_V2`?

File: streams/src/test/java/org/apache/kafka/streams/integration/EmitOnChangeIntegrationTest.java
Comment (wcarlson5) [reply streams/src/test/java/org/apache/kafka/streams/integration/EmitOnChangeIntegrationTest.java:237]: I will count using peek after the exception

File: streams/src/test/java/org/apache/kafka/streams/integration/EmitOnChangeIntegrationTest.java
Comment (wcarlson5) [reply streams/src/test/java/org/apache/kafka/streams/integration/EmitOnChangeIntegrationTest.java:245]: Yeah good idea. I will check to make sure it doesn't get processed more than once.

**Review by wcarlson5 — COMMENTED** [review]
@guozhangwang I understand the issue with performance and we wouldn't want to leave it that way for sure. But seeing that this is a strict improvement over the current behavior, as right now it won't make any progress in the case you describe, I don't think we need to hold off merging this. Also I think it should make writing tests for the backing off and reordering PR easier so it might make sense to merge it.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskExecutor.java:256]: For future reference, you can skip the `toString()` -- the logger should make the conversion implicitly

**Review by ableegoldman — APPROVED** [review]
Nice, LGTM. Will merge once the build passes

**Comment by guozhangwang** [comment]
> @guozhangwang I understand the issue with performance and we wouldn't want to leave it that way for sure. But seeing that this is a strict improvement over the current behavior, as right now it won't make any progress in the case you describe, I don't think we need to hold off merging this. Also I think it should make writing tests for the backing off and reordering PR easier so it might make sense to merge it.

To make sure that we do not introduce side effects beyond the named topology, what about we narrow the scope of this optimization to only named topology for now? Besides this, sounds reasonable to me.

**Pull request merged**
